src/datadog/remote_config/remote_config.cpp

```diff
@@ -231,41 +240,48 @@ void Manager::process_response(const nlohmann::json& json) {
         return;
       }
 
-      auto decoded_config =
-          base64_decode(target_it.value().at("raw").get<StringView>());
-
-      const auto config_json = nlohmann::json::parse(decoded_config);
+      auto raw_data = target_it->at("raw").get<StringView>();
+      auto decoded_config = base64_decode(raw_data);
 
       Configuration new_config;
-      new_config.id = config_json.at("id");
+      auto&& config_id = config_key_data->config_id;
+      new_config.id = std::string{config_id.data(), config_id.size()};
       new_config.path = std::string{config_path};
       new_config.hash = config_metadata.at("/hashes/sha256"_json_pointer);
       new_config.content = std::move(decoded_config);
-      new_config.version = config_json.at("revision");
-      new_config.product = *product;
-
-      const auto& targeted_service = config_json.at("service_target");
-      if (targeted_service.at("service").get<StringView>() !=
-              tracer_signature_.default_service ||
-          targeted_service.at("env").get<StringView>() !=
-              tracer_signature_.default_environment) {
-        new_config.state = Configuration::State::error;
-        new_config.error_message = "Wrong service targeted";
-      } else {
-        for (const auto& listener : listeners_per_product_[*product]) {
-          // Q: Two listeners on the same product. What should be the behaviour
-          // if one of the listeners report an error?
-          // R(@dmehala): Unspecified. For now, the config is marked with an
-          // error.
-          if (auto error_msg = listener->on_update(new_config)) {
-            new_config.state = Configuration::State::error;
-            new_config.error_message = std::move(*error_msg);
-          } else {
-            new_config.state = Configuration::State::acknowledged;
-          }
+      new_config.version = config_metadata.at("/custom/v"_json_pointer);
+      new_config.product = product;
+
+      // TODO: should be moved to the listener
+      if (product == product::Flag::APM_TRACING) {
+        const auto config_json = nlohmann::json::parse(new_config.content);
+        new_config.version = config_json.at("revision");
+
+        const auto& targeted_service = config_json.at("service_target");
+        if (targeted_service.at("service").get<StringView>() !=
+                tracer_signature_.default_service ||
+            targeted_service.at("env").get<StringView>() !=
+                tracer_signature_.default_environment) {
+          new_config.state = Configuration::State::error;
+          new_config.error_message = "Wrong service targeted";
+          goto skip_listeners;
+        }
+      }
```

Comment:
TODO.
[genuine question] Are you planning to move the code to the listener?

Reply:
[question] Do you mind if I push a commit to remove that technical debt?

Reply:
sure go ahead

Reply:
done in d11cd521edf61625bf49425de5ea0235a0b0b1d3